src/stdlib_experimental_quadrature.fypp

```diff
@@ -14,67 +15,130 @@ module stdlib_experimental_quadrature
+        ! "recursive" is an implementation detail
+        #:for k1, t1 in REAL_KINDS_TYPES
+        pure recursive module function simps_dx_${k1}$(y, dx, even) result(integral)
+            ${t1}$, dimension(:), intent(in) :: y
+            ${t1}$, intent(in) :: dx
             integer, intent(in), optional :: even
-            real(${KIND}$) :: integral
-        end function simps_dx_${KIND}$
+            ${t1}$ :: integral
+        end function simps_dx_${k1}$
         #:endfor
-        #:for KIND in REAL_KINDS
-        pure module function simps_x_${KIND}$(y, x, even) result(integral)
-            real(${KIND}$), dimension(:), intent(in) :: y
-            real(${KIND}$), dimension(size(y)), intent(in) :: x
+        #:for k1, t1 in REAL_KINDS_TYPES
+        pure recursive module function simps_x_${k1}$(y, x, even) result(integral)
+            ${t1}$, dimension(:), intent(in) :: y
+            ${t1}$, dimension(size(y)), intent(in) :: x
             integer, intent(in), optional :: even
-            real(${KIND}$) :: integral
-        end function simps_x_${KIND}$
+            ${t1}$ :: integral
+        end function simps_x_${k1}$
         #:endfor
     end interface simps
 
 
     interface simps_weights
-        #:for KIND in REAL_KINDS
-        pure module function simps_weights_${KIND}$(x, even) result(w)
-            real(${KIND}$), dimension(:), intent(in) :: x
-            real(${KIND}$), dimension(size(x)) :: w
+        #:for k1, t1 in REAL_KINDS_TYPES
+        pure recursive module function simps_weights_${k1}$(x, even) result(w)
+            ${t1}$, dimension(:), intent(in) :: x
             integer, intent(in), optional :: even
-        end function simps_weights_${KIND}$
+            ${t1}$, dimension(size(x)) :: w
+        end function simps_weights_${k1}$
         #:endfor
     end interface simps_weights
 
+
+    ! Interface for a simple f(x)-style integrand function.
+    ! Could become fancier as we learn about the performance
+    ! ramifications of different ways to do callbacks.
+    abstract interface
+        #:for k1, t1 in REAL_KINDS_TYPES
+        pure function integrand_${k1}$(x) result(f)
+            import :: ${k1}$
+            ${t1}$, intent(in) :: x
+            ${t1}$ :: f
+        end function integrand_${k1}$
+        #:endfor
+    end interface
+
+#:if QUAD_OK
+    ! Base class to avoid repeating kind parameter declaration.
+    type, abstract :: weight_t(kind)
+        integer, kind :: kind
+    end type weight_t
+
+    type, extends(weight_t) :: sin_weight_t
+        real(kind) :: omega
+    end type sin_weight_t
+
+    type, extends(weight_t) :: cos_weight_t
+        real(kind) :: omega
+    end type cos_weight_t
+
+    type, extends(weight_t) :: pole_weight_t
+        real(kind) :: c
+    end type pole_weight_t
+
+! gfortran 9.2.0 chokes on ICE if I include this ("buffer overflow detected")
+! Interestingly, though, the ICE happens while trying to build the trapz submodule
+! PDT bug?
```

Comment:
If you think it is a bug, did you report it?

Reply:
gfortran's PDT implementation is fundamentally broken and is known to need a rework. This is experimental code that should not be in the PR anyway.